### PR TITLE
EE-1100: Transfer error handling

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1100

- Propagates mint's ApiError to the caller through ext_ffi
- Replaces `Result<(), ()>` with proper error handling
- Augments Auction errors with different errors whenever transfer fails so it's not always `Transfer` but `Auction(something happened during transfer)`
- Adds assemblyscript support for said feature

Now `ApiError::Transfer` should be considered only for really unhandled transfer errors such as failed to call the contract, can't reconstruct `TransferredTo` etc. which should be rare cause.

### CODE REVIEW

**Review by Fraser999 — APPROVED** [review]
This looks good to me (bar a single nitpick).  However, I didn't thoroughly review the typescript changes, so it'd probably be worth waiting for a second reviewer before merging this.

File: smart_contracts/contract/src/contract_api/system.rs [hunk 1]
```diff
@@ -115,9 +117,16 @@ pub fn transfer_to_account(target: AccountHash, amount: U512, id: Option<u64>) -
             amount_size,
             id_ptr,
             id_size,
+            maybe_result_value.as_mut_ptr(),
         )
     };
-    TransferredTo::result_from(return_code)
+
+    // Propagate error (if there is)
```
Comment (EdHastingsCasperLabs) [inline smart_contracts/contract/src/contract_api/system.rs:124]: nitpick: // Propagate error (if any)

File: types/src/system_contract_errors/auction.rs [hunk 2]
```diff
@@ -100,11 +100,35 @@ pub enum Error {
...
+    #[fail(display = "Withdraw validator reward error")]
+    WithdrawValidatorReward = 31,
+    /// Failed to transfer desired amount into unbonding purse.
+    #[fail(display = "Transfer to unbonding purse error")]
+    TransferToUnbondingPurse = 32,
+
+    #[cfg(test)]
+    #[doc(hidden)]
+    #[fail(display = "Sentinel error")]
+    Sentinel,
```
Comment (EdHastingsCasperLabs) [inline types/src/system_contract_errors/auction.rs:126]: Why do we need this variant?

**Review by EdHastingsCasperLabs — APPROVED** [review]
see nitpicks

File: types/src/system_contract_errors/auction.rs (hunk 2 above)
Comment (mpapierski) [reply types/src/system_contract_errors/auction.rs:126]: Before this change

```
#[cfg(test)]
const MAX_ERROR_VALUE
```

was hardcoded and `enum Error` values grew, but tests didnt reflect that. With this `MAX_ERROR_VALUE` would be always highest value and thus all values will be tested (if we'd extend error enum in future)

**Pull request merged**